phase1-coordinator/src/storage/disk.rs

```diff
@@ -464,138 +376,181 @@ impl StorageLocator for Disk {
     }
 }
 
+pub struct DiskObjectReader {
+    _file: File,
+    memmap: memmap::Mmap,
+}
+
+impl Deref for DiskObjectReader {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &*self.memmap
+    }
+}
+
+impl AsRef<[u8]> for DiskObjectReader {
+    fn as_ref(&self) -> &[u8] {
+        self.memmap.as_ref()
+    }
+}
+
+impl ObjectReader for DiskObjectReader {}
+
+pub struct DiskObjectWriter {
+    _file: File,
+    memmap: memmap::MmapMut,
+}
+
+impl Deref for DiskObjectWriter {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &*self.memmap
+    }
+}
+
+impl DerefMut for DiskObjectWriter {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.memmap
+    }
+}
+
+impl AsMut<[u8]> for DiskObjectWriter {
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.memmap.as_mut()
+    }
+}
+
+impl ObjectWriter for DiskObjectWriter {
+    fn flush(&self) -> std::io::Result<()> {
+        self.memmap.flush()
+    }
+}
+
 impl StorageObject for Disk {
+    type Reader = DiskObjectReader;
+    type Writer = DiskObjectWriter;
+
     /// Returns an object reader for the given locator.
     #[inline]
-    fn reader<'a>(&self, locator: &Locator) -> Result<ObjectReader, CoordinatorError> {
+    fn reader<'a>(&self, locator: &Locator) -> Result<Self::Reader, CoordinatorError> {
+        let path = self.to_path(&locator)?;
+
         // Check that the locator exists in storage.
         if !self.exists(&locator) {
-            let locator = self.to_path(&locator)?;
-            error!("Locator {} missing in call to reader() in storage.", locator);
+            error!("Locator {} missing in call to reader() in storage.", path);
             return Err(CoordinatorError::StorageLocatorMissing);
         }
 
-        // Check that the given locator is opened in storage.
-        if !self.is_open(locator) {
-            error!("Locator in call to reader() is not opened in storage.");
-            return Err(CoordinatorError::StorageLocatorNotOpen);
-        }
+        let file = OpenOptions::new().read(true).open(path)?;
 
-        // Acquire the file read lock.
-        let reader = self
-            .open
-            .get(locator)
-            .ok_or(CoordinatorError::StorageLockFailed)?
-            .read()
-            .unwrap();
+        // Load the file into memory.
+        let memmap = unsafe { MmapOptions::new().map(&file)? };
```

Comment:
@julesdesmit this was the most important change for this PR. When `StorageObject` is obtained from `Disk` it will create the memmap, and when the `StorageObject` is dropped the memmap will also be destroyed. I was hoping that this would reduce the time that memmap was around in memory.